packages/node/src/utils/stellar.ts

```diff
@@ -0,0 +1,172 @@
+    .records;
+  const wrappedEffects: SorobanEffect[] = [];
+
+  effects.forEach((effect) => {
+    const wrappedEffect: SorobanEffect = {
+      ...effect,
+      ledger: null,
+      transaction: null,
+      operation: null,
+    };
+
+    wrappedEffects.push(wrappedEffect);
+  });
+
+  return wrappedEffects;
+}
+
+export async function fetchOperationsForTransaction(
+  transactionId: string,
+  api: Server,
+): Promise<SorobanOperation[]> {
+  const operations = (
+    await api.operations().forTransaction(transactionId).call()
+  ).records;
+
+  return Promise.all(
+    operations.map(async (op) => {
+      const wrappedOp: SorobanOperation = {
+        ...op,
+        ledger: null,
+        transaction: null,
+        effects: [] as SorobanEffect[],
+      };
+
+      const effects = (await fetchEffectsForOperation(op.id, api)).map(
+        (effect) => {
+          effect.operation = JSON.parse(JSON.stringify(wrappedOp));
+          return effect;
+        },
+      );
+
+      wrappedOp.effects.push(...effects);
+
+      return wrappedOp;
+    }),
+  );
+}
+
+export async function fetchTransactionsForLedger(
+  sequence: number,
+  api: Server,
+): Promise<SorobanTransaction[]> {
+  const transactions = (await api.transactions().forLedger(sequence).call())
+    .records;
+
+  return Promise.all(
+    transactions.map(async (tx) => {
+      let account: ServerApi.AccountRecord;
+      try {
+        account = await tx.account();
+      } catch (e) {
+        if ((e as Error).name === 'NotFoundError') {
+          account = null;
+        } else {
+          throw e;
+        }
+      }
+
+      const wrappedTx: SorobanTransaction = {
+        ...tx,
+        ledger: null,
+        account: account,
+        operations: [] as SorobanOperation[],
+        effects: [] as SorobanEffect[],
+      };
+
+      const operations = (await fetchOperationsForTransaction(tx.id, api)).map(
+        (op) => {
+          op.transaction = JSON.parse(JSON.stringify(wrappedTx));
+          op.effects = op.effects.map((effect) => {
+            effect.transaction = JSON.parse(JSON.stringify(wrappedTx));
+            return effect;
+          });
+          return op;
+        },
+      );
+
+      wrappedTx.operations.push(...operations);
+      operations.forEach((op) => {
+        wrappedTx.effects.push(...op.effects);
+      });
+
+      return wrappedTx;
+    }),
+  );
+}
+
+export async function fetchAndWrapLedger(
+  sequence: number,
+  api: Server,
+): Promise<SorobanBlock> {
+  const ledger = (await api
+    .ledgers()
+    .ledger(sequence)
+    .call()) as unknown as ServerApi.LedgerRecord;
+
+  const wrappedLedger: SorobanBlock = {
+    ...ledger,
+    transactions: [] as SorobanTransaction[],
+    operations: [] as SorobanOperation[],
+    effects: [] as SorobanEffect[],
+  };
+
+  const transactions = (await fetchTransactionsForLedger(sequence, api)).map(
+    (tx) => {
+      tx.ledger = JSON.parse(JSON.stringify(wrappedLedger));
```

Comment:
creating deep copies

Reply:
it needs to be cloned because when we wrap the tx to the ledger reference later, it leads to circular reference. so, we clone the ledger object to make sure tx.ledger.transaction remains null.